libibverbs/device.c

```diff
@@ -262,6 +267,8 @@ struct ibv_context *__ibv_open_device(struct ibv_device *device)
 	context->cmd_fd = cmd_fd;
 	pthread_mutex_init(&context->mutex, NULL);
 
+	ibverbs_device_hold(device);
+
```

Comment:
When the kernel recreats the /sys/.../uverbs0 it does so with a new directory inode (or it should). Since openat is inode based, the dirfd held open in libibverbs will continue to refer to the old, now empty,  directory from the original device, even though a path path based view will continue to see the new, replaced sysfs directory. The goal is not to block the kernel from replacing the directory, but to hold on to the original directory inode across the hotplug event. This allows a more transactional view of sysfs and is safe against hotplug - openat's on the dirfd will fail with ENOENT after a reuse event.